File: backend/app/schemas/relay.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List


class RelayRoomCreateRequest(BaseModel):
    mode: str = "single"
    difficulty: str = "normal"
    max_rounds: int = 20
    time_limit: int = 30
    max_players: int = 2
    password: Optional[str] = None


class RelayPlayerInfo(BaseModel):
    user_id: int
    username: str
    avatar_url: Optional[str] = None
    score: int = 0
    combo: int = 0
    is_host: bool = False

    class Config:
        from_attributes = True


class RelayRoundInfo(BaseModel):
    round: int
    user_id: int
    username: str
    verse: str
    poem_title: Optional[str] = None
    author: Optional[str] = None
    score: int = 0
    time_used: int = 0


class RelayRoomResponse(BaseModel):
    id: int
    room_code: str
    mode: str
    difficulty: str
    max_rounds: int
    time_limit: int
    status: str
    host_id: int
    host_username: str = ""
    max_players: int = 2
    has_password: bool = False
    created_at: datetime

    class Config:
        from_attributes = True


class RelayRoomDetailResponse(BaseModel):
    id: int
    room_code: str
    mode: str
    difficulty: str
    status: str
    current_round: int
    max_rounds: int
    time_limit: int
    next_char: Optional[str] = None
    players: List[RelayPlayerInfo] = []
    rounds: List[RelayRoundInfo] = []


class RelayStartResponse(BaseModel):
    room_id: int
    status: str
    current_round: int
    starter_verse: str
    starter_poem_title: str
    starter_author: str
    next_char: str
    started_at: datetime


class RelaySubmitRequest(BaseModel):
    verse: str


class RelaySubmitResponse(BaseModel):
    round: int
    verse: str
    poem_title: Optional[str] = None
    author: Optional[str] = None
    is_valid: bool
    match_type: str
    score: int
    next_char: str
    time_used: int
    combo: int


class RelayHintItem(BaseModel):
    verse: str
    poem_title: str
    author: str


class RelayHintResponse(BaseModel):
    hints: List[RelayHintItem]
    hint_count_used: int
    hint_count_max: int = 3


class RelayEndResultItem(BaseModel):
    user_id: int
    username: str
    total_score: int
    max_combo: int
    rounds_played: int
    avg_time: float
    rank: int


class RelayEndResponse(BaseModel):
    room_id: int
    total_rounds: int
    duration: int
    results: List[RelayEndResultItem]
    exp_gained: int
    points_gained: int
    new_achievements: List[str] = []


class RelayHistoryItem(BaseModel):
    room_id: int
    mode: str
    difficulty: str
    total_rounds: int
    total_score: int
    max_combo: int
    duration: int
    result: str
    played_at: datetime


class RelayHistoryResponse(BaseModel):
    items: List[RelayHistoryItem]
    total: int
    page: int
    page_size: int


class RelayRankingItem(BaseModel):
    rank: int
    user_id: int
    username: str
    avatar_url: Optional[str] = None
    total_score: int
    total_games: int
    max_combo: int
    best_rounds: int
    win_rate: float


class RelayRankingResponse(BaseModel):
    items: List[RelayRankingItem]
    total: int
    period: str


class RelayLobbyItem(BaseModel):
    id: int
    room_code: str
    difficulty: str
    max_rounds: int
    time_limit: int
    max_players: int = 2
    player_count: int = 0
    host_username: str = ""
    host_avatar: Optional[str] = None
    has_password: bool = False
    created_at: datetime


class RelayLobbyResponse(BaseModel):
    items: List[RelayLobbyItem]
    total: int
    page: int
    page_size: int

File: frontend/src/api/relay.ts
import request from '@/utils/request'

export interface RelayRoomCreateRequest {
  mode: string
  difficulty: string
  max_rounds: number
  time_limit: number
  max_players?: number
  password?: string
}

export interface RelayPlayerInfo {
  user_id: number
  username: string
  avatar_url: string | null
  score: number
  combo: number
  is_host: boolean
}

export interface RelayRoundInfo {
  round: number
  user_id: number
  username: string
  verse: string
  poem_title: string | null
  author: string | null
  score: number
  time_used: number
}

export interface RelayRoomResponse {
  id: number
  room_code: string
  mode: string
  difficulty: string
  max_rounds: number
  time_limit: number
  status: string
  host_id: number
  host_username: string
  max_players: number
  has_password: boolean
  created_at: string
}

export interface RelayLobbyItem {
  id: number
  room_code: string
  difficulty: string
  max_rounds: number
  time_limit: number
  max_players: number
  player_count: number
  host_username: string
  host_avatar: string | null
  has_password: boolean
  created_at: string
}

export interface RelayLobbyResponse {
  items: RelayLobbyItem[]
  total: number
  page: number
  page_size: number
}

export interface RelayRoomDetail {
  id: number
  room_code: string
  mode: string
  difficulty: string
  status: string
  current_round: number
  max_rounds: number
  time_limit: number
  next_char: string | null
  players: RelayPlayerInfo[]
  rounds: RelayRoundInfo[]
}

export interface RelayStartResponse {
  room_id: number
  status: string
  current_round: number
  starter_verse: string
  starter_poem_title: string
  starter_author: string
  next_char: string
  started_at: string
}

export interface RelaySubmitResponse {
  round: number
  verse: string
  poem_title: string | null
  author: string | null
  is_valid: boolean
  match_type: string
  score: number
  next_char: string
  time_used: number
  combo: number
}

export interface RelayHintItem {
  verse: string
  poem_title: string
  author: string
}

export interface RelayHintResponse {
  hints: RelayHintItem[]
  hint_count_used: number
  hint_count_max: number
}

export interface RelayEndResultItem {
  user_id: number
  username: string
  total_score: number
  max_combo: number
  rounds_played: number
  avg_time: number
  rank: number
}

export interface RelayEndResponse {
  room_id: number
  total_rounds: number
  duration: number
  results: RelayEndResultItem[]
  exp_gained: number
  points_gained: number
  new_achievements: string[]
}

export interface RelayHistoryItem {
  room_id: number
  mode: string
  difficulty: string
  total_rounds: number
  total_score: number
  max_combo: number
  duration: number
  result: string
  played_at: string
}

export interface RelayRankingItem {
  rank: number
  user_id: number
  username: string
  avatar_url: string | null
  total_score: number
  total_games: number
  max_combo: number
  best_rounds: number
  win_rate: number
}

export const relayApi = {
  async createRoom(data: RelayRoomCreateRequest): Promise<RelayRoomResponse> {
    const res: any = await request.post('/relay/rooms', data)
    return res.data || res
  },

  async joinRoom(roomCode: string, password?: string): Promise<any> {
    const params = password ? { password } : {}
    const res: any = await request.post(`/relay/rooms/${roomCode}/join`, null, { params })
    return res.data || res
  },

  async startGame(roomId: number): Promise<RelayStartResponse> {
    const res: any = await request.post(`/relay/rooms/${roomId}/start`)
    return res.data || res
  },

  async submitVerse(roomId: number, verse: string): Promise<RelaySubmitResponse> {
    const res: any = await request.post(`/relay/rooms/${roomId}/submit`, { verse })
    return res.data || res
  },

  async getRoomDetail(roomId: number): Promise<RelayRoomDetail> {
    const res: any = await request.get(`/relay/rooms/${roomId}`)
    return res.data || res
  },

  async getHint(roomId: number): Promise<RelayHintResponse> {
    const res: any = await request.get(`/relay/rooms/${roomId}/hint`)
    return res.data || res
  },

  async endGame(roomId: number): Promise<RelayEndResponse> {
    const res: any = await request.post(`/relay/rooms/${roomId}/end`)
    return res.data || res
  },

  async getHistory(params: { page?: number; page_size?: number } = {}): Promise<{ items: RelayHistoryItem[]; total: number; page: number; page_size: number }> {
    const res: any = await request.get('/relay/history', { params })
    return res.data || res
  },

  async getRankings(params: { period?: string; page?: number; page_size?: number } = {}): Promise<{ items: RelayRankingItem[]; total: number; period: string }> {
    const res: any = await request.get('/relay/rankings', { params })
    return res.data || res
  },

  async getLobby(params: { page?: number; page_size?: number } = {}): Promise<RelayLobbyResponse> {
    const res: any = await request.get('/relay/lobby', { params })
    return res.data || res
  },

  async quickMatch(): Promise<RelayRoomResponse> {
    const res: any = await request.post('/relay/quick-match')
    return res.data || res
  }
}
